### PR TITLE
32/64-bit time_t compatibility and RTC HAL refactoring

### PR DESCRIPTION
**Pull request by avtolstoy**
### Problem

`time_t` in newer versions of ARM GCC are now 64-bit, breaking dynalib ABI compatibility.

https://github.com/mirror/newlib-cygwin/commit/4de8754bac676fc7f29082ccf8281b1c73ce629e

In particular GCC 9.2.1 now breaks ABI compatibility in all the dynalib functions that use time_t or any compound types that contain `time_t` within them (e.g. `timeval`). 

### Solution

1. Provide necessary compaibility definitions/functions to keep 32-bit dynalib `time_t` functions intact for backwards compatibility
2. In order to simplify workarounds in (1), bump required ARM GCC version to 9.2.1, so that we don't have to support two sets of possible `time_t` sizes
3. RTC/External RTC HALs are somewhat refactored and cleaned up to increase time resolution (up to microseconds) on supported platforms
4. Fixes null pointer dereference in `spark_protocol_time_last_synced()` when called in `Spark_Handshake()`
5. Refactors wiring `Time` class to use reentrant versions of libc time functions: e.g. `localtime_r` vs `localtime`
6. Adds additional tests for RTC alarms and `time_t` compatibility changes

### Steps to Test

- `wiring/no_fixture`
- `wiring/sleep`, `wiring/sleep20`, `app/sleep_multiple_pins`
- `wiring/time_compat`

### Example App

N/A

### References

- [CH55923]

---

### Completeness

- [x] User is totes amazing for contributing!
- [x] Contributor has signed CLA ([Info here](https://github.com/spark/firmware/blob/develop/CONTRIBUTING.md))
- [x] Problem and Solution clearly stated
- [x] Run unit/integration/application tests on device
- [ ] Added documentation
- [ ] Added to CHANGELOG.md after merging (add links to docs and issues)

### CODE REVIEW

File: wiring/src/spark_wiring_time.cpp [hunk 1]
```diff
@@ -58,48 +58,16 @@ time_t dst_current_cache = 0;   // a cache of the DST offset currently being app
...
-
-/* Get converted Calendar time */
-/*
- static struct tm Get_CalendarTime(void)
-{
-	time_t unix_time = HAL_RTC_Get_UnixTime();
-	unix_time += time_zone_cache;
-	struct tm calendar_time = Convert_UnixTime_To_CalendarTime(unix_time);
+	struct tm calendar_time;
+	localtime_r(&unix_time, &calendar_time);
```
Comment (sergeuz) [inline wiring/src/spark_wiring_time.cpp:67]: > localtime_r

:+1:

File: hal/src/stm32f2xx/sleep_hal.cpp [hunk 2]
```diff
@@ -145,6 +145,12 @@ static int enterStopMode(const hal_sleep_config_t* config, hal_wakeup_source_bas
             extiInitStructure.EXTI_Trigger = EXTI_Trigger_Rising;
             extiInitStructure.EXTI_LineCmd = ENABLE;
             EXTI_Init(&extiInitStructure);
+
+            struct timeval tv = {
+                .tv_sec = seconds,
+                .tv_usec = 0
+            };
+            CHECK(hal_rtc_set_alarm(&tv, HAL_RTC_ALARM_FLAG_IN, nullptr, nullptr, nullptr));
```
[the remark this answers is not in the document]
Comment (avtolstoy) [reply hal/src/stm32f2xx/sleep_hal.cpp:153]: Yeah, it's alright: https://github.com/particle-iot/device-os/blob/fe6e0d91c935be9a4a1b188a445b821b6564055a/hal/src/stm32f2xx/sleep_hal.cpp#L236

Standby mode will cause a reset, so no handler will get executed anyway.

File: hal/shared/time_compat.h [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+ */
+
+#pragma once
+
+#include <sys/types.h>
+#include <sys/time.h>
+#include <stdint.h>
+#ifndef __cplusplus
+#include <assert.h>
+#endif // __cplusplus
```
[the remark this answers is not in the document]
Comment (avtolstoy) [reply hal/shared/time_compat.h:25]: Yeah, it doesn't:
```c++
#if __STDC_VERSION__ >= 201112L && !defined __cplusplus
# define static_assert _Static_assert
#endif
```

File: hal/network/lwip/socket_hal.cpp [hunk 4]
```diff
@@ -104,6 +131,19 @@ int sock_select(int nfds, fd_set* readfds, fd_set* writefds,
...
+                  fd_set* exceptfds, LIBC_TIMEVAL32* timeout) {
+  struct timeval tv = {};
+  if (timeout) {
+    tv.tv_sec = timeout->tv_sec;
+    tv.tv_usec = timeout->tv_usec;
+    return sock_select(nfds, readfds, writefds, exceptfds, &tv);
+  }
+  return sock_select(nfds, readfds, writefds, exceptfds, nullptr);
+}
+#endif // LIBC_64_BIT_TIME_T
```
[the remark this answers is not in the document]
Comment (avtolstoy) [reply hal/network/lwip/socket_hal.cpp:148]: Added an alias.

File: hal/network/lwip/socket_hal.cpp [hunk 5]
```diff
@@ -46,10 +46,37 @@ int sock_getsockname(int s, struct sockaddr* name, socklen_t* namelen) {
...
+  if (optname == SO_SNDTIMEO || optname == SO_RCVTIMEO) {
+    if (optlen && *optlen == sizeof(LIBC_TIMEVAL32) && optval) {
+      struct timeval tv = {};
+      socklen_t sz = sizeof(tv);
+      int r = lwip_getsockopt(s, level, optname, &tv, &sz);
+      if (!r) {
+        LIBC_TIMEVAL32* tv32 = (LIBC_TIMEVAL32*)optval;
+        tv32->tv_sec = tv.tv_sec;
+        tv32->tv_usec = tv.tv_usec;
+        return r;
```
[the remark this answers is not in the document]
Comment (avtolstoy) [reply hal/network/lwip/socket_hal.cpp]: Fixed.

File: hal/inc/rtc_hal.h [hunk 6]
```diff
@@ -42,13 +44,24 @@
...
-uint8_t HAL_RTC_Time_Is_Valid(void* reserved);
+typedef enum hal_rtc_alarm_flags {
+    HAL_RTC_ALARM_FLAG_IN = 0x01 // In provided amount of time, instead of an absolute timestamp
+} hal_rtc_alarm_flags;
+
+void hal_rtc_init(void);
+int hal_rtc_get_time(struct timeval* tv, void* reserved);
+int hal_rtc_set_time(const struct timeval* tv, void* reserved);
+bool hal_rtc_time_is_valid(void* reserved);
+int hal_rtc_set_alarm(const struct timeval* tv, uint32_t flags, hal_rtc_alarm_handler handler, void* context, void* reserved);
```
[the remark this answers is not in the document]
Comment (avtolstoy) [reply hal/inc/rtc_hal.h:58]: Added.

**Pull request merged**